src/mito2/src/memtable/merge_tree/merger.rs

```diff
@@ -0,0 +1,633 @@
+
+use std::cmp::{Ordering, Reverse};
+use std::collections::BinaryHeap;
+use std::fmt::Debug;
+use std::ops::Range;
+
+use datatypes::arrow::array::{
+    ArrayRef, TimestampMicrosecondArray, TimestampMillisecondArray, TimestampNanosecondArray,
+    TimestampSecondArray, UInt64Array,
+};
+use datatypes::arrow::datatypes::{DataType, TimeUnit};
+
+use crate::error::Result;
+use crate::memtable::merge_tree::data::{DataBatch, DataBufferReader, DataPartReader};
+use crate::memtable::merge_tree::PkIndex;
+
+pub trait Item: Clone {
+    type Key: Ord;
+
+    /// Remaining rows in item.
+    fn remaining(&self) -> usize;
+
+    /// Whether current item is exhausted.
+    fn is_empty(&self) -> bool;
+
+    /// The key range of item.
+    fn current_range(&self) -> Range<Self::Key>;
+
+    /// Searches given key in current item.
+    fn search_key(&self, key: &Self::Key) -> std::result::Result<usize, usize>;
+
+    /// Slice item.
+    fn slice(&self, range: Range<usize>) -> Self;
+}
+
+pub trait Node: Ord {
+    type Item: Item;
+
+    /// Returns current item of node and fetch next.
+    fn fetch_next(&mut self) -> Result<Self::Item>;
+
+    /// Returns true if current node is not exhausted.
+    fn is_valid(&self) -> bool;
+
+    /// Current item of node.
+    fn current_item(&self) -> &Self::Item;
+
+    /// Whether the other node is behind (exclusive) current node.
+    fn is_behind(&self, other: &Self) -> bool;
+
+    /// Skips first `num_to_skip` rows from node's current batch. If current batch is empty it fetches
+    /// next batch from the node.
+    ///
+    /// # Panics
+    /// If the node is EOF.
+    fn skip(&mut self, offset_to_skip: usize) -> Result<()>;
+}
+
+pub struct Merger<T, I> {
+    heap: BinaryHeap<T>,
+    current_item: Option<I>,
+}
+
+impl<T, I> Merger<T, I>
+where
+    I: Item + Debug,
+    T: Node<Item = I>,
+{
+    pub(crate) fn try_new(nodes: Vec<T>) -> Result<Self> {
+        let mut heap = BinaryHeap::with_capacity(nodes.len());
+        for node in nodes {
+            if node.is_valid() {
+                heap.push(node);
+            }
+        }
+        let mut merger = Merger {
+            heap,
+            current_item: None,
+        };
+        merger.next()?;
+        Ok(merger)
+    }
+
+    /// Returns true if current merger is still valid.
+    pub(crate) fn is_valid(&self) -> bool {
+        self.current_item.is_some()
+    }
+
+    /// Advances current merger to next item.
+    pub(crate) fn next(&mut self) -> Result<()> {
+        if let Some(mut top_node) = self.heap.pop() {
+            if let Some(next_node) = self.heap.peek() {
+                if next_node.is_behind(&top_node) {
+                    // does not overlap
+                    self.current_item = Some(top_node.fetch_next()?);
+                } else {
+                    let next_start = next_node.current_item().current_range().start;
+                    let res = match top_node.current_item().search_key(&next_start) {
```

Comment:
Could we define `search_key()` in the `Node` trait? Then we don't need to expose `current_range()` and `search_key()` in the `Item` trait.
```rust
match top_node.search_key(&next_node) {}
```

Reply:
> Could we define `search_key()` in the `Node` trait? Then we don't need to expose `current_range()` and `search_key()` in the `Item` trait.
> 
> ```rust
> match top_node.search_key(&next_node) {}
> ```

Then the meaning of `Node::search_key` would be ambiguous as it only searches the key in current item of node.

Reply:
Maybe rename it to `search_in_current_item`. I prefer to remove the associated type Key from the Item.

Reply:
I simplified the trait's associated types [here](https://github.com/GreptimeTeam/greptimedb/pull/3361/commits/0a0b55b57454d1869b57125d62729d219796a134).